assets/javascripts/tests.js

```diff
@@ -90,9 +90,27 @@ function renderTestName(data, type, row) {
         }
         html += dependencyHtml;
     }
-    if (row.comment_count) {
-        html += ' <a href="/tests/' + row.id + '#comments"><i class="test-label label_comment fa fa-comment" title="' +
-            row.comment_count + (row.comment_count != 1 ? ' comments' : ' comment') + ' available"' + '></i></a>';
+    if (row.comment_data) {
+        var bugs = row.comment_data.bugs;
+        bugs.forEach(function(bug) {
+            const css_class = bug.css_class;
+            const title = bug.title;
+            const url = bug.url;
+            html += '<span id="bug-' + bug.id + '"> ' + '<a href="' + htmlEscape(url) + '">';
+            html += '<i class="test-label ' + htmlEscape(css_class) + '" title="' + htmlEscape(title) + '"></i>';
+            html += '</a></span>';
+        });
+
+        if (row.comment_data.label) {
+            const label = row.comment_data.label;
+            html += '<span id="test-label-' + row.id + '">';
+            html += ' <i class="test-label label_' + htmlEscape(label) + ' fa fa-bookmark" title="Label: ' + htmlEscape(label) + '"></i>';
+            html += '</span>';
+        } else if (row.comment_data.comments) {
+            html += '<span id="comment-' + row.id + '"> ';
+            html += row.comment_data.comment_icon;
+            html += '</span>';
+        }
```

Comment:
This block looks like it's duplicating what we do as perl code in other cases. So you think it's feasible to use the same approach for deciding about which icon with which link we can add for each job result?

Reply:
> This block looks like it's duplicating what we do as perl code in other cases.

Yes, it'd duplicating some code in templates. Note that parts of it are already generated by perl helper functions.
An alternative would be to run the data through templates here as well and send the complete HTML for this column to Javascript.
Are we doing this already elsewhere?

> So you think it's feasible to use the same approach for deciding about which icon with which link we can add for each job result?

I don't understand this question.
I think it makes sense to use the same approach, because I assume the same result as in /tests/overview is wanted, so a *different* approach would be wrong.
Can you elaborate?

Reply:
Rendering this on the server likely slows down the loading time. I'd rather try to render everything in JavaScript. (It is not nice that we need to generate HTML code here. That's a limitation of the DataTables API as far as I know - so we have to live with it.)

Reply:
I think you already provided good suggestions. I would prefer if we use the same approach for both /tests as well as /tests/overview (maybe as well as "Next & previous job results"). Either that approach already exists for /tests/overview (maybe as well "Next & previous job results") or you need to create that new approach and use it everywhere so it will be the *same*

Reply:
Oh, I totally overlooked Next & previous!
Similar code is used there.
I only looked at /tests/overview.

Now, /tests/overview is generated server side via a template, /tests/list_ajax and job_next_previous_ajax are getting data and render HTML on client side.

What should I do about that (before implementing something and having to change it afterwards)?

Marius votes against rendering throgh a perl template in the Javascript cases.
I personally think that doesn't make much difference performance wise, most load comes from querieng the DB.

Oli requests that the same approach is used. That would mean either
1. changing /tests/overview to use Javascript instead or
2. changing both ajax routes to pregenerate HTML before sending it to Javascript

Reply:
Don't change it to pregenerated HTML. I've already converted several tables over time away from pregenerated HTML because it was too slow. (We have to generate the HTML for *all* rows on the server and then DataTables needs to parse all of this HTML at once and turn it into a functional DataTable.)

For this PR, I would at least share existing JavaScript code. So if we have already JavaScript code for rendering this for "Next & previous" we should reuse it. Streamlining with the test result overview is likely out of scope here.

Reply:
> We have to generate the HTML for _all_ rows on the server and then DataTables needs to parse all of this HTML

That's not true. I mean, I already *do* generate parts of the HTML here (calling the server side helper function `comment_icon` for example, which returns HTML.
What I meant was to just generate the HTML for the comment icons and do `html += comments`. Not the complete HTML table.

Reply:
Oh, you mean all rows instead of the only 10/25 that are rendered.
Sorry, I agree.

Reply:
Ah, so you would generate the HTML and pass it as JSON string. That should be better although it would still run the code for all rows on the server upfront. With client-side rendering we can just render the rows we need to render which just makes more sense from a scalability perspective. So I'd still try to unify the JavaScript code we have so far for now. If we'd ever modernize the frontend code (e.g. using some fancy new JavaScript framework) then we'd likely tend to render even more on the client. So moving this code to the server-side seems like a step backwards.